agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp

```diff
@@ -0,0 +1,246 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 MaxLinear
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "ieee802_3_link_metrics_collector.h"
+
+#include <bcl/network/network_utils.h>
+
+#include <easylogging++.h>
+
+#include <errno.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <net/if.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+// SPEED values
+#include <linux/ethtool.h>
+
+#define IFLIST_REPLY_BUFFER 8192
+
+namespace beerocks {
+
+/**
+ * @brief Netlink request type.
+ */
+struct nl_req_t {
+    /**
+     * Netlink message
+     */
+    struct nlmsghdr hdr;
+
+    /**
+     * "general form of address family dependent" message, i.e. how to tell which AF we are
+     * interested in. */
+    struct rtgenmsg gen;
+};
+
+static bool get_link_metrics(const struct nlmsghdr *h, const std::string &local_interface_name,
+                             sLinkMetrics &link_metrics)
+{
+    bool result = false;
+
+    struct ifinfomsg *iface = static_cast<ifinfomsg *>(NLMSG_DATA(h));
+    int length              = h->nlmsg_len - NLMSG_LENGTH(sizeof(*iface));
+
+    /**
+     * Loop over all attributes of the NEWLINK message
+     */
+    for (struct rtattr *attribute = IFLA_RTA(iface); RTA_OK(attribute, length);
+         attribute                = RTA_NEXT(attribute, length)) {
+        switch (attribute->rta_type) {
+        case IFLA_IFNAME:
+            /**
+             * This message contains the stats for the interface we are interested in
+             */
+            if (0 == strcmp(local_interface_name.c_str(), (char *)RTA_DATA(attribute))) {
+                result = true;
+            }
+            break;
+        case IFLA_STATS:
+            if (result) {
+                struct rtnl_link_stats *stats = (struct rtnl_link_stats *)RTA_DATA(attribute);
+
+                /**
+                 * Get interface speed and set PHY rate accordingly.
+                 * Speed values other than 100Mbps and 1Gbps are ignored.
+                 */
+                uint16_t phy_rate = UINT16_MAX;
+                uint32_t speed    = 0;
+                if (beerocks::net::network_utils::linux_iface_get_speed(local_interface_name,
+                                                                        speed)) {
+                    if (SPEED_100 == speed) {
+                        phy_rate = 100;
+                    } else if (SPEED_1000 == speed) {
+                        phy_rate = 1000;
+                    }
+                }
+
+                link_metrics.transmitter.packet_errors       = stats->tx_errors;
```

Comment:
Sorry but I cannot see the timestamp field you mentioned. Maybe I'm not looking at the right document.  Can you please clarify which are those R1 and R2 docs?
I agree with you that if we cannot include the time measurement window along with the counter, then it is useless to provide the number of packets/errors **during** such measurement period. 

I think however that, even if we could provide timestamp or elapsed time since last measure, we have no way to know if the number of packets we read with the netlink socket have wrapped (and how many times) since last call to get link metrics. Although such wrapping is quite improbable to go unnoticed, given the frequency with which calls to get link metrics will be probably issued, we cannot guarantee that wrapping won't happen. Keeping timestamps is useless if they are not provided by the kernel.

Reply:
Just to be clear: since the code is there now anyway, I would merge it as is. But when later we refactor with #900, I think we should also look into using rtnl_link instead of doing all the parsing manually. However, before going into that, we should check if the libnl-tiny used in openwrt also has it.

Reply:
Read in https://wireless.wiki.kernel.org/en/developers/documentation/nl80211

> libnl tiny
> OpenWrt folks created a tiny version of libnl based on a git snapshot, which only contains genl, not rtnetlink or any of the netfilter stuff, and compiles down to less than 30k in binary size. You can find it here:
> https://git.openwrt.org/?p=openwrt/openwrt.git;a=tree;f=package/libs/libnl-tiny;hb=HEAD